application_command.go

```diff
@@ -0,0 +1,193 @@
+package disgord
+
+import (
+	"context"
+	"fmt"
+	"net/http"
+
+	"github.com/andersfylling/disgord/internal/httd"
+)
+
+type ApplicationCommandType int
+
+const (
+	_ ApplicationCommandType = iota
+	ApplicationCommandChatInput
+	ApplicationCommandUser
+	ApplicationCommandMessage
+)
+
+type ApplicationCommandOptionChoice struct {
+	Name  string      `json:"name"`
+	Value interface{} `json:"value"`
+}
+
+type ApplicationCommandOption struct {
+	Type        OptionType                        `json:"type"`
+	Name        string                            `json:"name"`
+	Description string                            `json:"description"`
+	Required    bool                              `json:"required"`
+	Choices     []*ApplicationCommandOptionChoice `json:"choices"`
+	Options     []*ApplicationCommandOption       `json:"options"`
+}
+
+type ApplicationCommandDataOption struct {
+	Name    string                          `json:"name"`
+	Type    OptionType                      `json:"type"`
+	Value   interface{}                     `json:"value"`
+	Options []*ApplicationCommandDataOption `json:"options"`
+}
+
+type ApplicationCommandPermissionType int
+
+const (
+	_ ApplicationCommandPermissionType = iota
+	ApplicationCommandPermissionRole
+	ApplicationCommandPermissionUser
+)
+
+type ApplicationCommandPermissions struct {
+	ID         Snowflake                        `json:"id"`
+	Type       ApplicationCommandPermissionType `json:"type"`
+	Permission bool                             `json:"permission"`
+}
+type GuildApplicationCommandPermissions struct {
+	ID            Snowflake                        `json:"id"`
+	ApplicationID Snowflake                        `json:"application_id"`
+	GuildID       Snowflake                        `json:"guild_id"`
+	Permissions   []*ApplicationCommandPermissions `json:"permissions"`
+}
+
+type ApplicationCommand struct {
+	ID                Snowflake                   `json:"id"`
+	Type              ApplicationCommandType      `json:"type"`
+	ApplicationID     Snowflake                   `json:"application_id"`
+	GuildID           Snowflake                   `json:"guild_id"`
+	Name              string                      `json:"name"`
+	Description       string                      `json:"description"`
+	Options           []*ApplicationCommandOption `json:"options"`
+	DefaultPermission bool                        `json:"default_permission,omitempty"`
+}
+
+type UpdateApplicationCommand struct {
+	Name              string                      `json:"name,omitempty"`
+	DefaultPermission bool                        `json:"default_permission,omitempty"`
```

Comment:
What if I want to update the name of a Application command which has default permissions?
If I send
```go
&UpdateApplicationCommand{ Name: "test" }
```

it's going to overwrite the permissions and set it to false. I think these needs to be pointers.

Reply:
see https://github.com/andersfylling/disgord/pull/420

Reply:
this is also relevant: https://github.com/andersfylling/disgord/pull/432

but the naming convention you've used fits greatly.